_posts/2017-10-24-Bun.md

```diff
@@ -0,0 +1,61 @@
+---
+title: Fuzzing for CI Workflows
+layout: post
+author: yomimono
+category: General
+tags: ocaml fuzzing afl crowbar travis ci platform tooling testing
+---
+
+# `ocaml-bun`: Fuzzing for CI
+
+Stephen Dolan recently presented [crowbar](https://github.com/stedolan/crowbar) at the 2017 OCaml Workshop. Crowbar bridges a gap between property-based testing frameworks and instrumentation-based automated testing techniques.  Tests written in Crowbar can be executed by the wildly popular and successful [American Fuzzy Lop](http://lcamtuf.coredump.cx/afl) fuzzer.
+
+American Fuzzy Lop provides a command-line tool for running tests, `afl-fuzz`.  `afl-fuzz` will invoke one test instance to run a given test continuously on different inputs, keeping track of which inputs generated crashes (or, for Crowbar, were counterexamples for a given proposition).  The default settings for `afl-fuzz` don't terminate when crashes are discovered, or when new execution paths haven't been found.  The user is expected to terminate all the `afl-fuzz` instances themselves when they're satisfied that enough work has been done (as indicated, colorfully, in the excellent GUI), and then inspect the files left by `afl-fuzz` during the run.
+
+The workflow is a great fit for a security researcher trying to find vulnerabilities in existing, released software (especially one who wants to assign work to all of the CPUs they have access to explicitly), but it doesn't fit well into a continuous testing workflow at all.  `ocaml-bun` is a tool for running Crowbar tests in your CI environment, so fuzzing is as easy as unit testing -- both to perform, and to inform decisions with.
+
+## Goals for `ocaml-bun`
+
+* Deliver "counterexample found, here it is" or "no counterexamples found in the allotted time" in automated run contexts.
+* Require no persistent storage for the user to be able to analyze and reproduce crash inputs and execution.
+* Make full use of available resources both in shared cloud environments and self-hosted test environments.
+* Play nicely with container testing environments.
+
+## What's `ocaml-bun` do so far?
```

Comment:
This section is very clear, but we are missing a very simple usage example of bun before. Perhaps a synthetic example with a few lines of `ml` with an obvious bug that it can find?

Reply:
Would a link out to the demo in Crowbar itself be OK?  I don't want to confuse people by linking the actual test code that you write with the fuzzer orchestrator you use to manage fuzzing it.

Reply:
A link to the demo would work :)